v3/lints/rfc/lint_crl_revoked_certificates_field_empty.go

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+	"golang.org/x/crypto/cryptobyte"
+	cryptobyte_asn1 "golang.org/x/crypto/cryptobyte/asn1"
+)
+
+type revokedCertificates struct{}
+
+/*
+RFC 5280: 5.1.2.6
+
+	When there are no revoked certificates, the revoked certificates list
+	MUST be absent.
+*/
+func init() {
+	lint.RegisterRevocationListLint(&lint.RevocationListLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_crl_revoked_certificates_field_must_be_empty",
+			Description:   "When the revokedCertificates field is empty, it MUST be absent from the DER-encoded ASN.1 data structure.",
+			Citation:      "RFC 5280: 5.1.2.6",
+			Source:        lint.RFC5280,
+			EffectiveDate: util.RFC5280Date,
+		},
+		Lint: NewEmptyRevokedCertificates,
+	})
+}
+
+func NewEmptyRevokedCertificates() lint.RevocationListLintInterface {
+	return &revokedCertificates{}
+}
+
+func (l *revokedCertificates) CheckApplies(c *x509.RevocationList) bool {
+	// This lint is to verify that the TBSCertList.revokedCertificates field,
+	// when empty, is indeed missing from the DER-encoded ASN.1 bytes.
+	if c != nil && len(c.RevokedCertificates) == 0 {
+		return true
+	}
+
+	return false
+}
+
+func (l *revokedCertificates) Execute(c *x509.RevocationList) *lint.LintResult {
+	// This is a modified version of x509.ParseRevocationList that extracts the
+	// raw DER-encoded bytes that comprise a CRL and parses away layers until
+	// the optional `revokedCertificates` field of a TBSCertList is either found
+	// or confirmed to be missing from the ASN.1 data structure.
+	input := cryptobyte.String(c.Raw)
+
+	// From crypto/x509/parser.go: we read the SEQUENCE including length and tag
+	// bytes so that we can populate RevocationList.Raw, before unwrapping the
+	// SEQUENCE so it can be operated on
+	if !input.ReadASN1Element(&input, cryptobyte_asn1.SEQUENCE) {
+		return &lint.LintResult{Status: lint.Fatal, Details: "malformed CRL"}
+	}
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return &lint.LintResult{Status: lint.Fatal, Details: "malformed CRL"}
+	}
+
+	var tbs cryptobyte.String
+	// From crypto/x509/parser.go: do the same trick again as above to extract
+	// the raw bytes for Certificate.RawTBSCertificate
+	if !input.ReadASN1Element(&tbs, cryptobyte_asn1.SEQUENCE) {
+		return &lint.LintResult{Status: lint.Fatal, Details: "malformed TBS CRL"}
+	}
+	if !tbs.ReadASN1(&tbs, cryptobyte_asn1.SEQUENCE) {
+		return &lint.LintResult{Status: lint.Fatal, Details: "malformed TBS CRL"}
+	}
+
+	// Skip optional version
+	tbs.SkipOptionalASN1(cryptobyte_asn1.INTEGER)
```

Comment:
Is there a chance that the CRL is so malformed that these fields don't exist? What happens in that case?

Reply:
For what it's worth, [Boulder]https://github.com/letsencrypt/boulder/tree/2b8dea9821233bf4dd185e69f322f20abb8f2017/linter/lints/rfc) contains a lint that looks for a missing `issuer` field.